data/src/main/java/com/linkedin/data/DataList.java

```diff
@@ -239,6 +272,67 @@ public int dataComplexHashCode()
     return _dataComplexHashCode;
   }
 
+  public void traverse(Data.TraverseCallback callback, Data.CycleChecker cycleChecker) throws IOException
+  {
+    if (isEmpty())
+    {
+      callback.emptyList();
+    }
+    else
+    {
+      try
+      {
+        cycleChecker.startList(this);
+        callback.startList(this);
+
+        // Use Java 8 forEach to minimize intermediary object creation for better performance.
+        try
+        {
+          //
+          // If this is backed by a specific list, delegate traversal to the specific list. Else, use Java 8 forEach
+          // to avoid intermediary object creation.
+          //
+          if (_encapsulatesSpecificList)
+          {
+            ((SpecificList) _list).traverse(callback, cycleChecker);
+          }
+          else {
+            final int[] index = {0};
```

Comment:
This is actually faster per JMH, because .get(i) results in a range check everytime which foreach helps avoid.
The array allocation is on the heap but short lived, so it doesn't impact GC much.